contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,794 @@
+we will need to create one new attribute in the NodeStatus structure:
+  * `DevCapacity`: Describing the device capacity of the node
+
+```golang
+type NodeStatus struct {
+	DevCapacity []Device
+}
+```
+
+We also introduce the `Devices` field in the Containers status so that user
+can know what devices were assigned to the container.
+
+```golang
+type ContainerStatus struct {
+	Devices []Device
+}
+```
+
+Note that we will be using OIR to schedule and trigger the device plugin
+in parallel.
+So when a Device plugin registers two `foo-device` the node status will be
+updated to advertise 2 `extensions.kubernetes.io/foo-device`.
+
+If a user wants to trigger the device plugin he only needs to request this
+OIR in his Pod Spec.
+
+## Upgrading your cluster
+
+TLDR: If you are upgrading either Kubelet or any device plugin the safest way
+is to drain the node of all pods and upgrade.
+However depending on what you are upgrading and what changes happened then it
+is completely possible to only restart just Kubelet or just the device plugin.
+
+### Upgrading Kubelet
+
+This assumes that the Device Plugins running on the nodes fully implement the
+protocol and are able to recover from a Kubelet crash.
+
+Then, as long as the Device Plugin API does not change upgrading Kubelet can be done
+seamlessly through a Kubelet restart.
+
+However, as mentioned in the Versioning section, we currently expect the Device
+Plugin's API version to match exactly the Kubelet's Device Plugin API version.
+
+Therefore if the Device Plugin API version change then you will have to change
+the Device Plugin too.
+Consider draining the node in that case.
+
+When the Device Plugin API becomes a stable feature, versionning should be
+backward compatible and even if Kubelet has a different Device Plugin API,
+it should not require a Device Plugin upgrade.
+
+### Upgrading Device Plugins
+
+Because we cannot enforce what the different Device Plugins will do, we cannot
+say for certain that upgrading a device plugin will not crash any containers
+on the node.
+
+It is therefore up to the Device Plugin vendors to specify if the Device Plugins
+can be upgraded without impacting any running containers.
+
+As mentioned earlier, the safest way is to drain the node before upgrading
+the Device Plugins.
+
+## Difference Between Implementations
+
+The main difference between implementation 1 and 2 are:
+* This implementation allows vendors to run device specific code before
+  starting the containers requesting these devices.
+* This implementation allows users to know what devices were assigned
+  to a container
+* This implementation does not need checkpointing
+* This implementation has a clear separation of concerns, every functions
+  does one thing and only one. Every actor has only one explicit role:
+  * Kubelet's gRPC is in charge of keeping track of Device Plugins
+  * The Device Plugin's gRPC is in charge of handling devices
+
+
+
+# Proposed Implementation 2
```

Comment:
If you are mentioning this, do mention an author and may be call it as `Alternatives Considered`.